packages/android_alarm_manager/android/src/main/java/io/flutter/plugins/androidalarmmanager/AndroidAlarmManagerPlugin.java

```diff
@@ -16,25 +17,76 @@
+ *   any given Dart callback from that point forward. Thus, at this moment the plugin is
+ *   fully initialized and ready to execute arbitrary Dart tasks in the background. The Dart
+ *   side of this plugin sends the Android side a "AlarmService.initialized" message to
+ *   signify that the Dart is ready to execute tasks.</li>
+ * </ol>
+ */
 public class AndroidAlarmManagerPlugin implements MethodCallHandler, ViewDestroyListener {
-  /** Plugin registration. */
+  /**
+   * Registers this plugin with an associated Flutter execution context, represented by
+   * the given {@link Registrar}.
+   * <p>
+   * Once this method is executed, an instance of {@code AndroidAlarmManagerPlugin}
+   * will be connected to, and running against, the associated Flutter execution context.
+   */
   public static void registerWith(Registrar registrar) {
-    final MethodChannel channel =
+    // alarmManagerPluginChannel is the channel responsible for receiving the following messages
+    // from the main Flutter app:
+    // - "AlarmService.start"
+    // - "Alarm.oneShot"
+    // - "Alarm.periodic"
+    // - "Alarm.cancel"
+    final MethodChannel alarmManagerPluginChannel =
         new MethodChannel(
             registrar.messenger(),
             "plugins.flutter.io/android_alarm_manager",
             JSONMethodCodec.INSTANCE);
-    final MethodChannel backgroundChannel =
+
+    // backgroundCallbackChannel is the channel responsible for receiving the following messages
+    // from the background isolate that was setup by this plugin:
+    // - "AlarmService.initialized"
+    //
+    // This channel is also responsible for sending requests from Android to Dart to execute Dart
+    // callbacks in the background isolate. Those messages are sent with an empty method name because
+    // they are the only messages that this channel sends to Dart.
+    final MethodChannel backgroundCallbackChannel =
         new MethodChannel(
             registrar.messenger(),
             "plugins.flutter.io/android_alarm_manager_background",
             JSONMethodCodec.INSTANCE);
+
+    // Instantiate a new AndroidAlarmManagerPlugin, connect the primary and background
+    // method channels for Android/Flutter communication, and listen for FlutterView
+    // destruction so that this plugin can move itself to background mode.
     AndroidAlarmManagerPlugin plugin = new AndroidAlarmManagerPlugin(registrar.context());
-    channel.setMethodCallHandler(plugin);
-    backgroundChannel.setMethodCallHandler(plugin);
+    alarmManagerPluginChannel.setMethodCallHandler(plugin);
+    backgroundCallbackChannel.setMethodCallHandler(plugin);
     registrar.addViewDestroyListener(plugin);
-    AlarmService.setBackgroundChannel(backgroundChannel);
+
+    // The AlarmService expects to hold a static reference to the plugin's background
+    // method channel.
+    // TODO(mattcarroll): this static reference implies that only one instance of this plugin
```

Comment:
It's unlikely that this happens with alarm manager, but some plugins may offer background and foreground execution. In which case this will break, isn't it?

Reply:
Unfortunately, I think the entire concept of background vs foreground is a bit ambiguous at the moment. But to your specific point, can you provide an example what you think would cause a breakage here?